File: zinc-flow-java/src/main/java/zincflow/processors/RouteOnAttribute.java
package zincflow.processors;

import zincflow.core.FlowFile;
import zincflow.core.Processor;
import zincflow.core.ProcessorResult;

import java.util.ArrayList;
import java.util.List;
import java.util.Locale;

/// Routes a FlowFile to a named relationship based on the first matching
/// rule in a rules-spec string. Rules form:
/// {@code "routeA: attr OP value; routeB: attr OP value"}.
///
/// Supported operators (mirrors zinc-flow-csharp):
/// <ul>
///   <li>{@code ==} / {@code EQ}, {@code !=} / {@code NEQ} — equality</li>
///   <li>{@code CONTAINS}, {@code STARTSWITH}, {@code ENDSWITH} — substring tests</li>
///   <li>{@code MATCHES} — full-string regex</li>
///   <li>{@code EXISTS} — attribute presence (value ignored)</li>
///   <li>{@code GT}, {@code GE}, {@code LT}, {@code LE} — numeric comparison
///       (both sides parsed as doubles; fall back to lexicographic for non-numeric values)</li>
/// </ul>
///
/// No rule matches → routes to "unmatched". Downstream connections
/// decide what happens from there; wire a terminal processor or an
/// UpdateAttribute to "unmatched" to handle the fall-through explicitly.
public final class RouteOnAttribute implements Processor {

    private final List<Rule> rules;

    public RouteOnAttribute(String spec) {
        this.rules = parse(spec);
    }

    @Override
    public ProcessorResult process(FlowFile ff) {
        for (Rule rule : rules) {
            if (rule.evaluate(ff.attributes())) {
                return ProcessorResult.routed(rule.name(), ff);
            }
        }
        return ProcessorResult.routed("unmatched", ff);
    }

    public List<Rule> rules() { return rules; }

    // --- Parsing ---

    private static List<Rule> parse(String spec) {
        if (spec == null || spec.isBlank()) return List.of();
        List<Rule> out = new ArrayList<>();
        String[] entries = spec.split(";");
        for (int i = 0; i < entries.length; i++) {
            String entry = entries[i].trim();
            if (entry.isEmpty()) continue;
            int colonIdx = entry.indexOf(':');
            if (colonIdx <= 0) {
                throw new IllegalArgumentException(
                        "RouteOnAttribute: malformed route at index " + i + ": '" + entry
                                + "' — expected 'name: attr OP value'");
            }
            String routeName = entry.substring(0, colonIdx).trim();
            String condition = entry.substring(colonIdx + 1).trim();
            // EXISTS takes no value: "routeA: attr EXISTS" is valid.
            String[] parts = condition.split("\\s+", 3);
            if (parts.length < 2) {
                throw new IllegalArgumentException(
                        "RouteOnAttribute: route '" + routeName + "' has malformed condition: '"
                                + condition + "' — expected 'attr OP [value]'");
            }
            Op op = Op.parse(parts[1], routeName);
            String value = parts.length >= 3 ? parts[2] : "";
            if (op != Op.EXISTS && parts.length < 3) {
                throw new IllegalArgumentException(
                        "RouteOnAttribute: route '" + routeName + "' operator " + op
                                + " requires a value but none was provided");
            }
            out.add(new Rule(routeName, parts[0], op, value));
        }
        return List.copyOf(out);
    }

    public enum Op {
        EQ, NEQ, CONTAINS, STARTSWITH, ENDSWITH, MATCHES, EXISTS, GT, GE, LT, LE;

        static Op parse(String token, String routeName) {
            return switch (token.toUpperCase(Locale.ROOT)) {
                case "==", "EQ" -> EQ;
                case "!=", "NEQ" -> NEQ;
                case "CONTAINS" -> CONTAINS;
                case "STARTSWITH" -> STARTSWITH;
                case "ENDSWITH" -> ENDSWITH;
                case "MATCHES" -> MATCHES;
                case "EXISTS" -> EXISTS;
                case ">", "GT" -> GT;
                case ">=", "GE" -> GE;
                case "<", "LT" -> LT;
                case "<=", "LE" -> LE;
                default -> throw new IllegalArgumentException(
                        "RouteOnAttribute: route '" + routeName + "' has unsupported operator '"
                                + token + "' — valid: EQ/==, NEQ/!=, CONTAINS, STARTSWITH, ENDSWITH, MATCHES, EXISTS, GT/>, GE/>=, LT/<, LE/<=");
            };
        }
    }

    public record Rule(String name, String attribute, Op op, String value) {
        boolean evaluate(java.util.Map<String, String> attributes) {
            if (op == Op.EXISTS) {
                return attributes.containsKey(attribute);
            }
            String actual = attributes.get(attribute);
            if (actual == null) return false;
            return switch (op) {
                case EQ -> value.equals(actual);
                case NEQ -> !value.equals(actual);
                case CONTAINS -> actual.contains(value);
                case STARTSWITH -> actual.startsWith(value);
                case ENDSWITH -> actual.endsWith(value);
                case MATCHES -> actual.matches(value);
                case GT, GE, LT, LE -> compareTo(actual, value, op);
                case EXISTS -> true; // unreachable — handled above
            };
        }

        /// Numeric comparison when both sides parse as doubles, lexicographic
        /// fallback otherwise. Matches the zinc-flow-csharp semantics: string
        /// compare when values aren't parseable as numbers, so GT/LT can be
        /// used on version strings, timestamps-as-ISO-strings, etc.
        private static boolean compareTo(String actual, String expected, Op op) {
            int cmp;
            try {
                double a = Double.parseDouble(actual);
                double b = Double.parseDouble(expected);
                cmp = Double.compare(a, b);
            } catch (NumberFormatException ignored) {
                cmp = actual.compareTo(expected);
            }
            return switch (op) {
                case GT -> cmp > 0;
                case GE -> cmp >= 0;
                case LT -> cmp < 0;
                case LE -> cmp <= 0;
                default -> false; // unreachable
            };
        }
    }
}

File: zinc-flow-go/zinc-out/processors/route_on_attribute.go
package processors

import (
	"fmt"
	"zinc-flow/core"
	"strings"
)

//line /home/vrjoshi/proj/caravan-flow/zinc-flow-go/src/processors/route_on_attribute.zn:25
type Route struct {
	Relationship string
	Attribute string
	Op string
	Value string
}

func NewRoute(relationship string, attribute string, op string, value string) Route {
	return Route{Relationship: relationship, Attribute: attribute, Op: op, Value: value}
}

func (s Route) String() string {
	return fmt.Sprintf("Route(relationship=%v, attribute=%v, op=%v, value=%v)", s.Relationship, s.Attribute, s.Op, s.Value)
}

//line /home/vrjoshi/proj/caravan-flow/zinc-flow-go/src/processors/route_on_attribute.zn:27
type RouteOnAttribute struct {
	routes []Route
}

func NewRouteOnAttribute(routesSpec string) *RouteOnAttribute {
	s := &RouteOnAttribute{routes: []Route{}}
	s.parseRoutes(routesSpec)
	return s
}

func (s *RouteOnAttribute) parseRoutes(spec string) {
	if spec == "" {
		return
	}
	entries := splitAndTrim(spec, ";")
	for _, entry := range entries {
		colonIdx := indexOfByte(entry, ":")
		if colonIdx <= 0 {
			continue
		}
		name := trimSpace(entry[0:colonIdx])
		conditionStr := trimSpace(entry[colonIdx + 1:len(entry)])
		parts := splitOnSpace(conditionStr, 3)
		if len(parts) < 2 {
			continue
		}
		attribute := parts[0]
		op := upperCase(parts[1])
		value := ""
		if len(parts) >= 3 {
			value = parts[2]
		}
		s.routes = append(s.routes, NewRoute(name, attribute, op, value))
	}
}

func (s *RouteOnAttribute) Process(ff core.FlowFile) core.ProcessorResult {
	for _, route := range s.routes {
		if evaluate(route, ff.Attributes) {
			return core.NewRouted(route.Relationship, ff)
		}
	}
	return core.NewRouted("unmatched", ff)
}


//line /home/vrjoshi/proj/caravan-flow/zinc-flow-go/src/processors/route_on_attribute.zn:75
func RouteOnAttributeFactory(ctx *core.ScopedContext, config map[string]string) core.ProcessorFn {
	spec := ""
	if func() bool { _, _ok := config["routes"]; return _ok }() {
		spec = config["routes"]
	}
	return NewRouteOnAttribute(spec)
}

//line /home/vrjoshi/proj/caravan-flow/zinc-flow-go/src/processors/route_on_attribute.zn:85
func evaluate(r Route, attrs map[string]string) bool {
	if r.Op == "EXISTS" {
		return func() bool { _, _ok := attrs[r.Attribute]; return _ok }()
	}
	if !func() bool { _, _ok := attrs[r.Attribute]; return _ok }() {
		return false
	}
	actual := attrs[r.Attribute]
	if r.Op == "EQ" {
		return actual == r.Value
	}
	if r.Op == "NEQ" {
		return actual != r.Value
	}
	if r.Op == "CONTAINS" {
		return strings.Contains(actual, r.Value)
	}
	if r.Op == "STARTSWITH" {
		return strings.HasPrefix(actual, r.Value)
	}
	if r.Op == "ENDSWITH" {
		return strings.HasSuffix(actual, r.Value)
	}
	if r.Op == "GT" {
		return actual > r.Value
	}
	if r.Op == "LT" {
		return actual < r.Value
	}
	return false
}

//line /home/vrjoshi/proj/caravan-flow/zinc-flow-go/src/processors/route_on_attribute.zn:119
func splitAndTrim(s string, sep string) []string {
	out := []string{}
	start := 0
	i := 0
	for i < len(s) {
		if s[i:i + 1] == sep {
			part := trimSpace(s[start:i])
			if part != "" {
				out = append(out, part)
			}
			start = i + 1
		}
		i = i + 1
	}
	tail := trimSpace(s[start:len(s)])
	if tail != "" {
		out = append(out, tail)
	}
	return out
}

//line /home/vrjoshi/proj/caravan-flow/zinc-flow-go/src/processors/route_on_attribute.zn:140
func indexOfByte(s string, b string) int {
	i := 0
	for i < len(s) {
		if s[i:i + 1] == b {
			return i
		}
		i = i + 1
	}
	return -1
}

//line /home/vrjoshi/proj/caravan-flow/zinc-flow-go/src/processors/route_on_attribute.zn:151
func trimSpace(s string) string {
	start := 0
	for start < len(s) {
		c := s[start:start + 1]
		if c != " " && c != "\t" {
			break
		}
		start = start + 1
	}
	stop := len(s)
	for stop > start {
		c := s[stop - 1:stop]
		if c != " " && c != "\t" {
			break
		}
		stop = stop - 1
	}
	return s[start:stop]
}

//line /home/vrjoshi/proj/caravan-flow/zinc-flow-go/src/processors/route_on_attribute.zn:171
func splitOnSpace(s string, maxParts int) []string {
	out := []string{}
	start := 0
	i := 0
	for i < len(s) && len(out) < maxParts - 1 {
		c := s[i:i + 1]
		if c == " " || c == "\t" {
			if i > start {
				out = append(out, s[start:i])
			}
			for i < len(s) && (s[i:i + 1] == " " || s[i:i + 1] == "\t") {
				i = i + 1
			}
			start = i
			continue
		}
		i = i + 1
	}
	if start < len(s) {
		out = append(out, s[start:len(s)])
	}
	return out
}

//line /home/vrjoshi/proj/caravan-flow/zinc-flow-go/src/processors/route_on_attribute.zn:196
func upperCase(s string) string {
	out := ""
	i := 0
	for i < len(s) {
		c := s[i:i + 1]
		if c >= "a" && c <= "z" {
			b := byte(s[i])
			out = out + string(byte(b - 32))
		} else {
			out = out + c
		}
		i = i + 1
	}
	return out
}

